File: MQTT/src/1/publisher_and_subscriber.cpp
// MQTT C/C++ Example using Eclipse Paho MQTT C Library
// Compile: gcc -o mqtt_example mqtt_example.c -lpaho-mqtt3c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "MQTTClient.h"

// Configuration constants
#define BROKER_ADDRESS "tcp://localhost:1883"
#define CLIENT_ID "CPP_Example_Client"
#define TOPIC "home/sensor/temperature"
#define QOS 1
#define TIMEOUT 10000L

// Callback when connection is lost
void connlost(void *context, char *cause) {
    printf("\nConnection lost\n");
    printf("Cause: %s\n", cause);
}

// Callback when message arrives
int msgarrvd(void *context, char *topicName, int topicLen, MQTTClient_message *message) {
    printf("Message arrived on topic: %s\n", topicName);
    printf("Message: %.*s\n", message->payloadlen, (char*)message->payload);
    printf("QoS: %d\n", message->qos);
    printf("Retained: %s\n\n", message->retained ? "yes" : "no");
    
    MQTTClient_freeMessage(&message);
    MQTTClient_free(topicName);
    return 1;
}

// Callback when delivery is complete
void delivered(void *context, MQTTClient_deliveryToken dt) {
    printf("Message delivery confirmed (token: %d)\n", dt);
}

// Publisher function
int mqtt_publish_example() {
    MQTTClient client;
    MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    MQTTClient_message pubmsg = MQTTClient_message_initializer;
    MQTTClient_deliveryToken token;
    int rc;

    // Create MQTT client
    MQTTClient_create(&client, BROKER_ADDRESS, CLIENT_ID,
                      MQTTCLIENT_PERSISTENCE_NONE, NULL);
    
    // Set connection options
    conn_opts.keepAliveInterval = 20;
    conn_opts.cleansession = 1;
    
    // Set Last Will and Testament
    MQTTClient_willOptions will_opts = MQTTClient_willOptions_initializer;
    will_opts.topicName = "home/sensor/status";
    will_opts.message = "offline";
    will_opts.retained = 1;
    will_opts.qos = 1;
    conn_opts.will = &will_opts;

    // Connect to broker
    if ((rc = MQTTClient_connect(client, &conn_opts)) != MQTTCLIENT_SUCCESS) {
        printf("Failed to connect, return code %d\n", rc);
        return rc;
    }
    printf("Connected to broker successfully\n");

    // Publish messages
    for (int i = 0; i < 5; i++) {
        char payload[50];
        sprintf(payload, "Temperature: %.1f°C", 20.0 + i * 0.5);
        
        pubmsg.payload = payload;
        pubmsg.payloadlen = strlen(payload);
        pubmsg.qos = QOS;
        pubmsg.retained = 0;
        
        MQTTClient_publishMessage(client, TOPIC, &pubmsg, &token);
        printf("Publishing: %s\n", payload);
        
        rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
        printf("Message delivered (token: %d)\n\n", token);
        
        sleep(1);
    }

    // Disconnect
    MQTTClient_disconnect(client, 10000);
    MQTTClient_destroy(&client);
    return rc;
}

// Subscriber function
int mqtt_subscribe_example() {
    MQTTClient client;
    MQTTClient_connectOptions conn_opts = MQTTClient_connectOptions_initializer;
    int rc;

    // Create client
    MQTTClient_create(&client, BROKER_ADDRESS, "CPP_Subscriber",
                      MQTTCLIENT_PERSISTENCE_NONE, NULL);
    
    // Set callbacks
    MQTTClient_setCallbacks(client, NULL, connlost, msgarrvd, delivered);

    // Connection options
    conn_opts.keepAliveInterval = 20;
    conn_opts.cleansession = 1;

    // Connect
    if ((rc = MQTTClient_connect(client, &conn_opts)) != MQTTCLIENT_SUCCESS) {
        printf("Failed to connect, return code %d\n", rc);
        return rc;
    }
    printf("Subscriber connected to broker\n");

    // Subscribe to topic with wildcards
    printf("Subscribing to topic: home/sensor/#\n");
    MQTTClient_subscribe(client, "home/sensor/#", QOS);

    // Keep listening for messages
    printf("Waiting for messages...\n");
    for (int i = 0; i < 30; i++) {
        sleep(1);
    }

    // Unsubscribe and disconnect
    MQTTClient_unsubscribe(client, "home/sensor/#");
    MQTTClient_disconnect(client, 10000);
    MQTTClient_destroy(&client);
    return rc;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: %s [pub|sub]\n", argv[0]);
        return 1;
    }

    if (strcmp(argv[1], "pub") == 0) {
        return mqtt_publish_example();
    } else if (strcmp(argv[1], "sub") == 0) {
        return mqtt_subscribe_example();
    } else {
        printf("Invalid argument. Use 'pub' or 'sub'\n");
        return 1;
    }
}

File: MQTT/src/1/publisher_and_subscriber.rs
// MQTT Rust Example using rumqttc library
// Add to Cargo.toml:
// [dependencies]
// rumqttc = "0.23"
// tokio = { version = "1", features = ["full"] }

use rumqttc::{AsyncClient, MqttOptions, QoS, Event, Packet, LastWill};
use std::time::Duration;
use tokio::time;

// Publisher example with various MQTT features
async fn mqtt_publisher() -> Result<(), Box<dyn std::error::Error>> {
    // Configure MQTT options
    let mut mqttoptions = MqttOptions::new("rust_publisher", "localhost", 1883);
    mqttoptions.set_keep_alive(Duration::from_secs(20));
    mqttoptions.set_clean_session(true);
    
    // Set Last Will and Testament
    let lwt = LastWill::new(
        "home/sensor/status",
        "offline",
        QoS::AtLeastOnce,
        true, // retained
    );
    mqttoptions.set_last_will(lwt);

    // Create async client and event loop
    let (client, mut eventloop) = AsyncClient::new(mqttoptions, 10);
    
    // Spawn task to handle events
    tokio::spawn(async move {
        loop {
            match eventloop.poll().await {
                Ok(Event::Incoming(Packet::ConnAck(_))) => {
                    println!("Publisher connected to broker");
                }
                Ok(Event::Incoming(Packet::PubAck(ack))) => {
                    println!("Message acknowledged: pkid={}", ack.pkid);
                }
                Ok(Event::Outgoing(_)) => {},
                Ok(_) => {},
                Err(e) => {
                    eprintln!("Connection error: {:?}", e);
                    time::sleep(Duration::from_secs(1)).await;
                }
            }
        }
    });

    // Wait for connection
    time::sleep(Duration::from_millis(500)).await;

    // Publish messages with different QoS levels
    println!("Publishing messages...\n");
    
    for i in 0..5 {
        let topic = "home/sensor/temperature";
        let payload = format!("Temperature: {:.1}°C", 20.0 + i as f32 * 0.5);
        
        // QoS 0 - Fire and forget
        client.publish(topic, QoS::AtMostOnce, false, payload.clone()).await?;
        println!("Published (QoS 0): {}", payload);
        
        time::sleep(Duration::from_millis(500)).await;
    }

    // Publish with QoS 1 - At least once delivery
    let payload = "Critical: Temperature spike detected!";
    client.publish(
        "home/sensor/alerts",
        QoS::AtLeastOnce,
        false,
        payload
    ).await?;
    println!("\nPublished alert (QoS 1): {}", payload);

    // Publish retained message - stays on broker
    let status = "online";
    client.publish(
        "home/sensor/status",
        QoS::AtLeastOnce,
        true, // retained
        status
    ).await?;
    println!("Published retained status: {}\n", status);

    // Wait for acknowledgments
    time::sleep(Duration::from_secs(2)).await;

    // Disconnect gracefully
    client.disconnect().await?;
    println!("Publisher disconnected");
    
    Ok(())
}

// Subscriber example with pattern matching
async fn mqtt_subscriber() -> Result<(), Box<dyn std::error::Error>> {
    // Configure MQTT options
    let mut mqttoptions = MqttOptions::new("rust_subscriber", "localhost", 1883);
    mqttoptions.set_keep_alive(Duration::from_secs(20));
    
    // Use persistent session to queue messages when offline
    mqttoptions.set_clean_session(false);

    // Create client and event loop
    let (client, mut eventloop) = AsyncClient::new(mqttoptions, 10);

    // Subscribe to multiple topics with different QoS levels
    client.subscribe("home/sensor/#", QoS::AtLeastOnce).await?;
    client.subscribe("home/+/temperature", QoS::AtMostOnce).await?;
    client.subscribe("home/sensor/alerts", QoS::ExactlyOnce).await?;
    
    println!("Subscriber connected and subscribed to topics");
    println!("Subscriptions:");
    println!("  - home/sensor/# (QoS 1)");
    println!("  - home/+/temperature (QoS 0)");
    println!("  - home/sensor/alerts (QoS 2)");
    println!("\nWaiting for messages...\n");

    // Event loop to receive messages
    loop {
        match eventloop.poll().await {
            Ok(Event::Incoming(Packet::Publish(publish))) => {
                let topic = publish.topic.clone();
                let payload = String::from_utf8_lossy(&publish.payload);
                let qos = publish.qos;
                let retained = publish.retain;
                
                println!("─────────────────────────────────");
                println!("Topic:    {}", topic);
                println!("Payload:  {}", payload);
                println!("QoS:      {:?}", qos);
                println!("Retained: {}", retained);
                println!("─────────────────────────────────\n");
            }
            Ok(Event::Incoming(Packet::ConnAck(connack))) => {
                println!("Connected! Session present: {}\n", connack.session_present);
            }
            Ok(Event::Incoming(Packet::SubAck(suback))) => {
                println!("Subscription confirmed: pkid={:?}\n", suback.pkid);
            }
            Ok(Event::Outgoing(_)) => {},
            Ok(_) => {},
            Err(e) => {
                eprintln!("Error: {:?}", e);
                time::sleep(Duration::from_secs(1)).await;
            }
        }
    }
}

// Advanced example: Request-Response pattern using MQTT
async fn mqtt_request_response() -> Result<(), Box<dyn std::error::Error>> {
    let mut mqttoptions = MqttOptions::new("rpc_client", "localhost", 1883);
    mqttoptions.set_keep_alive(Duration::from_secs(20));

    let (client, mut eventloop) = AsyncClient::new(mqttoptions, 10);
    
    // Subscribe to response topic
    let response_topic = "home/rpc/response";
    client.subscribe(response_topic, QoS::AtLeastOnce).await?;

    // Spawn event handler
    tokio::spawn(async move {
        loop {
            if let Ok(Event::Incoming(Packet::Publish(p))) = eventloop.poll().await {
                let payload = String::from_utf8_lossy(&p.payload);
                println!("RPC Response: {}", payload);
            }
        }
    });

    time::sleep(Duration::from_millis(500)).await;

    // Send RPC request
    let request = r#"{"method": "getData", "params": {"sensor": "temp01"}}"#;
    client.publish(
        "home/rpc/request",
        QoS::AtLeastOnce,
        false,
        request
    ).await?;
    
    println!("RPC Request sent: {}", request);
    time::sleep(Duration::from_secs(3)).await;

    Ok(())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = std::env::args().collect();
    
    if args.len() < 2 {
        println!("Usage: {} [pub|sub|rpc]", args[0]);
        return Ok(());
    }

    match args[1].as_str() {
        "pub" => mqtt_publisher().await?,
        "sub" => mqtt_subscriber().await?,
        "rpc" => mqtt_request_response().await?,
        _ => println!("Invalid argument. Use 'pub', 'sub', or 'rpc'"),
    }

    Ok(())
}